### DISCUSSION TITLE
Making visible signature fields required

### DISCUSSION THREAD

**Discussion by stefangassner** [MatthiasValvekens/pyHanko#280] [Q&A] [answered]
Hi, 

thank you for this great library!

I would like to know if there is a way to set a visible signature field to "required"?
I'm creating a pdf using wkhtmltopdf and add the signature fields afterwards to the pdf.
Afterwards InSign is used as the signature process but only required fields are considered.

I could not find anything to add this attribute.

thanks for your help!

Cheers

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi @stefangassner,


Odd. First time I'm hearing of a batch processing tool that actually looks at those attributes.

Anyway, there's indeed no convenient API to do this using pyHanko's field setup API, but there's a relatively straightforward workaround. If `w` is the PDF writer object in which you just created a field, then this should do the trick in most (=pretty much all) cases. Warning: untested code.

```python
field_dict = next(f for f in w.root['/AcroForm']['/Fields'] if f['/T'] == your_field_name)
field_dict['/Ff'] = NumberObject(2)
```

If you've got other stuff in `/Ff`, replace the assignment with a bitwise OR. For further info see Table 227 in ISO 32000-2:2020 (you can get a free copy of the standard from pdfa.org these days).

**Comment by stefangassner** [comment]
Hi @MatthiasValvekens,

thank you for your response.

Your snippet helped me for a start and I ended up using this code:

```
writer = IncrementalPdfFileWriter(pdf_bytes)
for signature_field in signature_fields:
    append_signature_field(writer, SigFieldSpec(sig_field_name=signature_field['signature_name'],
                                                        on_page=signature_field['page'], box=signature_field['box']))
for i, field in enumerate(signature_fields):
    writer.root['/AcroForm']['/Fields'][i].update({NameObject('/Ff'): NumberObject(2)})
writer.write_in_place()
```

Thank you!
